File: dto/hub.go
package dto

// PublishHubDirectoryReq 发布目录到 Hub 请求
type PublishHubDirectoryReq struct {
	APIKey               string            `json:"api_key"`                // API Key（私有化部署需要）
	SourceUser           string            `json:"source_user"`            // 源用户
	SourceApp            string            `json:"source_app"`             // 源应用
	SourceDirectoryPath  string            `json:"source_directory_path"`  // 源目录完整路径
	Name                 string            `json:"name"`                  // 目录名称
	Description          string            `json:"description"`            // 目录描述
	Category             string            `json:"category"`               // 分类
	Tags                 []string          `json:"tags"`                   // 标签
	ServiceFeePersonal   float64           `json:"service_fee_personal"`   // 个人用户服务费
	ServiceFeeEnterprise float64           `json:"service_fee_enterprise"` // 企业用户服务费
	Version              string            `json:"version"`                // 版本号（默认 v1）
	DirectoryTree        *DirectoryTreeNode `json:"directory_tree"`        // 目录树结构（递归，支持嵌套）
}

// DirectoryTreeNode 目录树节点（用于发布目录，包含文件内容）
type DirectoryTreeNode struct {
	Name           string              `json:"name"`            // 目录名称
	Path           string              `json:"path"`            // 目录完整路径
	Files          []*FileSnapshotInfo `json:"files"`           // 该目录下的文件列表（包含内容）
	Subdirectories []*DirectoryTreeNode `json:"subdirectories"` // 子目录列表（递归）
}

// FileSnapshotInfo 文件快照信息
type FileSnapshotInfo struct {
	FileName     string `json:"file_name"`     // 文件名（不含 .go 后缀）
	RelativePath string `json:"relative_path"` // 文件相对路径
	Content      string `json:"content"`       // 文件代码内容
	FileType     string `json:"file_type"`     // 文件类型
	FileVersion  string `json:"file_version"`  // 文件版本号
}

// PublishHubDirectoryResp 发布目录到 Hub 响应
type PublishHubDirectoryResp struct {
	HubDirectoryID  int64  `json:"hub_directory_id"`
	HubDirectoryURL string `json:"hub_directory_url"`
	DirectoryCount  int    `json:"directory_count"` // 包含的子目录数量
	FileCount       int    `json:"file_count"`      // 包含的文件数量
}

// UpdateHubDirectoryReq 更新目录到 Hub 请求（用于 push）
type UpdateHubDirectoryReq struct {
	APIKey               string            `json:"api_key"`                // API Key（私有化部署需要）
	HubDirectoryID       int64             `json:"hub_directory_id"`       // Hub 目录 ID（必需）
	SourceDirectoryPath  string            `json:"source_directory_path"`   // 源目录完整路径
	Name                 string            `json:"name"`                   // 目录名称（可选）
	Description          string            `json:"description"`            // 目录描述（可选）
	Category             string            `json:"category"`               // 分类（可选）
	Tags                 []string          `json:"tags"`                  // 标签（可选）
	ServiceFeePersonal   float64           `json:"service_fee_personal"`   // 个人用户服务费（可选）
	ServiceFeeEnterprise float64           `json:"service_fee_enterprise"` // 企业用户服务费（可选）
	Version              string            `json:"version"`                // 新版本号（必需）
	DirectoryTree        *DirectoryTreeNode `json:"directory_tree"`        // 目录树结构
}

// UpdateHubDirectoryResp 更新目录到 Hub 响应
type UpdateHubDirectoryResp struct {
	HubDirectoryID  int64  `json:"hub_directory_id"`
	HubDirectoryURL string `json:"hub_directory_url"`
	DirectoryCount  int    `json:"directory_count"` // 包含的子目录数量
	FileCount       int    `json:"file_count"`      // 包含的文件数量
	OldVersion      string `json:"old_version"`     // 旧版本号
	NewVersion      string `json:"new_version"`     // 新版本号
}

// HubDirectoryListResp Hub 目录列表响应
type HubDirectoryListResp struct {
	Items    []*HubDirectoryDTO `json:"items"`
	Page     int                 `json:"page"`
	PageSize int                 `json:"page_size"`
	Total    int64               `json:"total"`
}

// HubDirectoryDTO Hub 目录 DTO（用于 API 返回）
type HubDirectoryDTO struct {
	ID        int64   `json:"id"`
	CreatedAt string  `json:"created_at"`
	UpdatedAt string  `json:"updated_at"`

	// 基本信息
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Category    string   `json:"category"`
	Tags        []string `json:"tags"` // 标签数组

	// 目录路径信息
	PackagePath  string `json:"package_path"`  // 目录路径
	FullCodePath string `json:"full_code_path"` // 完整代码路径

	// 源信息
	SourceUser          string `json:"source_user"`
	SourceApp           string `json:"source_app"`
	SourceDirectoryPath string `json:"source_directory_path"`

	// 发布信息
	PublisherUsername string `json:"publisher_username"`
	PublishedAt       string `json:"published_at"`

	// 服务费信息
	ServiceFeePersonal   float64 `json:"service_fee_personal"`
	ServiceFeeEnterprise float64 `json:"service_fee_enterprise"`

	// 统计信息
	DownloadCount int     `json:"download_count"`
	TrialCount    int     `json:"trial_count"`
	Rating        float64 `json:"rating"`

	// 版本信息
	Version    string `json:"version"`     // 版本号（如 v1.0.0）
	VersionNum int    `json:"version_num"` // 版本号（数字部分）

	// 统计信息（快照）
	DirectoryCount int `json:"directory_count"` // 子目录数量
	FileCount      int `json:"file_count"`      // 文件数量
	FunctionCount  int `json:"function_count"`  // 函数数量
}

// HubDirectoryDetailDetailResp Hub 目录详情响应
type HubDirectoryDetailDetailResp struct {
	HubDirectoryDTO
	DirectoryTree *DirectoryTreeNode  `json:"directory_tree,omitempty"` // 目录树结构（可选，Files 字段可能为空，仅用于展示）
	Files         []*DirectoryFileDTO `json:"files,omitempty"`         // 文件列表（可选）
}

// DirectoryFileDTO 目录文件 DTO
type DirectoryFileDTO struct {
	FileName     string `json:"file_name"`     // 文件名
	RelativePath string `json:"relative_path"` // 文件相对路径
	FileType     string `json:"file_type"`     // 文件类型
	FileSize     int    `json:"file_size"`     // 文件大小
	// 注意：content 不包含在列表中，需要单独获取
}

File: enterprise_impl/hub/frontend/src/api/hub.ts
/**
 * Hub API 客户端
 * 
 * Hub 是 AI-Agent-OS 的应用市场，提供目录发布、浏览、克隆等功能。
 */

import { getHubBaseURL } from '@/config/hub'
import { get, post, del } from '@/utils/request'

/** Hub 前端配置（主站前端地址，用于「试用」跳转） */
export interface HubConfigResp {
  main_site_url: string  // 主站前端地址（如 http://localhost:5173 或 http://125.122.96.207:8999），不是后端 API
}

/**
 * 获取 Hub 配置（主站前端地址）
 * 用于点击「试用」时跳转：主站前端 + /workspace + 目录 full_code_path
 */
export function getHubConfig(): Promise<HubConfigResp> {
  const baseURL = getHubBaseURL()
  return get<HubConfigResp>(`${baseURL}/config`)
}

/**
 * Hub 目录信息
 */
export interface HubDirectoryInfo {
  id: number
  created_at: string
  updated_at: string
  status?: string                 // active=在架；deleted=已下架（列表不展示，通过链接仍可访问）
  name: string
  description: string
  category: string
  tags: string[]
  package_path: string              // 目录路径
  full_code_path: string            // 完整代码路径
  source_user: string
  source_app: string
  source_directory_path: string
  publisher_username: string
  published_at: string
  service_fee_personal: number
  service_fee_enterprise: number
  download_count: number
  trial_count: number
  rating: number
  version: string
  directory_count: number           // 子目录数量
  file_count: number                // 文件数量
  function_count: number            // 函数数量
  copy_url?: string                 // 复制链接，格式 hub://host/path@version，用于粘贴到工作空间
  star_count?: number               // 星星数（类似 GitHub star）
  has_starred?: boolean              // 当前用户是否已加星（仅详情返回，未登录或未加星为 false）
}

/**
 * 获取 Hub 目录列表响应
 */
export interface HubDirectoryListResp {
  items: HubDirectoryInfo[]
  page: number
  page_size: number
  total: number
}

/** 费用筛选：全部 / 免费 / 收费 */
export type FeeTypeFilter = '' | 'free' | 'paid'

/**
 * 获取 Hub 目录列表
 */
/** 列表排序：latest=最新，hot=热门，stars=按星数，downloads=按复制数 */
export type OrderByFilter = 'latest' | 'hot' | 'stars' | 'downloads'

export async function getHubDirectoryList(params?: {
  page?: number
  page_size?: number
  search?: string
  category?: string
  publisher_username?: string
  /** 只看自己：true 时后端按当前用户过滤（需带 token/网关带 X-Request-User），未登录返回 401 */
  mine_only?: boolean
  fee_type?: FeeTypeFilter
  order_by?: OrderByFilter
}): Promise<HubDirectoryListResp> {
  const baseURL = getHubBaseURL()
  const url = `${baseURL}/directories`
  const p = { ...params }
  if (p.fee_type === '') delete (p as Record<string, unknown>).fee_type
  if (p.mine_only === false) delete (p as Record<string, unknown>).mine_only
  return get<HubDirectoryListResp>(url, p || {})
}

/**
 * 目录文件信息
 */
export interface DirectoryFileInfo {
  file_name: string
  relative_path: string
  file_type: string
  file_size: number
}

/**
 * 函数信息
 */
export interface HubFunctionInfo {
  id: number
  name: string
  code: string
  full_code_path: string
  description: string
  template_type: string
  tags: string[]
  ref_id: number
  version: string
  version_num: number
}

/**
 * 目录树节点
 */
export interface DirectoryTreeNode {
  type: 'package' | 'function'  // 节点类型：package（目录）或 function（函数）
  name: string
  path: string
  // ⭐ files 字段已移除，不再返回和展示文件
  functions?: HubFunctionInfo[]  // 函数列表
  subdirectories: DirectoryTreeNode[]
}

/**
 * Hub 目录详情
 */
export interface HubDirectoryDetail extends HubDirectoryInfo {
  directory_tree?: DirectoryTreeNode  // 目录树结构（可选）
  version_description?: string        // 当前查看版本的更新说明（推送时填的「本版本更新说明」）
  // ⭐ files 字段已移除，不再返回和展示文件
}

/**
 * 目录版本项（历史版本列表）
 */
export interface HubDirectoryVersionItem {
  version: string
  version_num: number
  snapshot_at: string
  is_current: boolean
  description?: string               // 本版本更新说明（可选）
  publisher_username?: string       // 该版本的上传人
}

export interface DownloadHubDirectoryBundleParams {
  hubDirectoryId?: number
  fullCodePath?: string
  version?: string
}

export interface DownloadHubDirectoryBundleResult {
  blob: Blob
  filename: string
}

function parseDownloadFileName(contentDisposition: string | null): string {
  if (!contentDisposition) {
    return 'hub-directory-bundle.json'
  }

  const utf8Match = contentDisposition.match(/filename\*=UTF-8''([^;]+)/i)
  if (utf8Match?.[1]) {
    return decodeURIComponent(utf8Match[1])
  }

  const plainMatch = contentDisposition.match(/filename="([^"]+)"/i) || contentDisposition.match(/filename=([^;]+)/i)
  if (plainMatch?.[1]) {
    return plainMatch[1].trim()
  }

  return 'hub-directory-bundle.json'
}

async function parseDownloadError(response: Response): Promise<string> {
  try {
    const data = await response.json() as { msg?: string; message?: string }
    return data.msg || data.message || '导出失败'
  } catch {
    return '导出失败'
  }
}

/**
 * 获取 Hub 目录详情
 * @param version 可选，不传则返回最新版本
 */
/** 用 full_code_path 获取详情（推荐，用于详情页 URL 展示） */
export async function getHubDirectoryDetailByPath(
  fullCodePath: string,
  includeTree?: boolean,
  version?: string
): Promise<HubDirectoryDetail> {
  const baseURL = getHubBaseURL()
  const url = `${baseURL}/directories/detail`
  const params: Record<string, string | number | boolean> = {
    full_code_path: fullCodePath,
    include_tree: includeTree ?? false
  }
  if (version) params.version = version
  return get<HubDirectoryDetail>(url, params)
}

/** 用 hub_directory_id 获取详情（兼容旧链接） */
export async function getHubDirectoryDetail(
  hubDirectoryId: number,
  includeTree?: boolean,
  version?: string
): Promise<HubDirectoryDetail> {
  const baseURL = getHubBaseURL()
  const url = `${baseURL}/directories/detail`
  const params: Record<string, string | number | boolean> = {
    hub_directory_id: hubDirectoryId,
    include_tree: includeTree ?? false
  }
  if (version) params.version = version
  return get<HubDirectoryDetail>(url, params)
}

/**
 * 获取 Hub 目录版本列表（用于详情页右侧历史版本）
 */
/** 用 full_code_path 获取版本列表（推荐） */
export async function getHubDirectoryVersionsByPath(
  fullCodePath: string
): Promise<{ items: HubDirectoryVersionItem[] }> {
  const baseURL = getHubBaseURL()
  const url = `${baseURL}/directories/versions`
  return get<{ items: HubDirectoryVersionItem[] }>(url, {
    full_code_path: fullCodePath
  })
}

export async function getHubDirectoryVersions(
  hubDirectoryId: number
): Promise<{ items: HubDirectoryVersionItem[] }> {
  const baseURL = getHubBaseURL()
  const url = `${baseURL}/directories/versions`
  return get<{ items: HubDirectoryVersionItem[] }>(url, {
    hub_directory_id: hubDirectoryId
  })
}

export async function downloadHubDirectoryBundle(
  params: DownloadHubDirectoryBundleParams
): Promise<DownloadHubDirectoryBundleResult> {
  const baseURL = getHubBaseURL()
  const query = new URLSearchParams()
  if (params.hubDirectoryId) {
    query.set('hub_directory_id', String(params.hubDirectoryId))
  }
  if (params.fullCodePath) {
    query.set('full_code_path', params.fullCodePath)
  }
  if (params.version) {
    query.set('version', params.version)
  }

  const token = localStorage.getItem('token') || ''
  const response = await fetch(`${baseURL}/directories/export_bundle?${query.toString()}`, {
    headers: token.trim()
      ? { 'X-Token': token }
      : undefined
  })

  const contentDisposition = response.headers.get('content-disposition')
  if (!contentDisposition?.toLowerCase().includes('attachment')) {
    throw new Error(await parseDownloadError(response))
  }

  return {
    blob: await response.blob(),
    filename: parseDownloadFileName(contentDisposition)
  }
}

/** 为目录加星（需要登录） */
export async function starHubDirectory(hubDirectoryId: number): Promise<void> {
  const baseURL = getHubBaseURL()
  const url = `${baseURL}/directories/${hubDirectoryId}/star`
  await post(url, {})
}

/** 取消星星（需要登录） */
export async function unstarHubDirectory(hubDirectoryId: number): Promise<void> {
  const baseURL = getHubBaseURL()
  const url = `${baseURL}/directories/${hubDirectoryId}/star`
  await del(url)
}

/** 删除应用（软删除：只改状态，数据保留，通过链接仍可访问；仅发布者可操作，需要登录） */
export async function deleteHubDirectory(hubDirectoryId: number): Promise<void> {
  const baseURL = getHubBaseURL()
  const url = `${baseURL}/directories/${hubDirectoryId}`
  await del(url)
}

// ==================== Pub Key 管理 ====================

export interface PubKeyItem {
  id: number
  name: string
  key_prefix: string
  created_at: string
  last_used_at: string | null
}

export interface GeneratePubKeyResp {
  id: number
  name: string
  key: string
  key_prefix: string
}

export async function generatePubKey(name: string): Promise<GeneratePubKeyResp> {
  const baseURL = getHubBaseURL()
  return post<GeneratePubKeyResp>(`${baseURL}/pub_key/generate`, { name })
}

export async function listPubKeys(): Promise<PubKeyItem[]> {
  const baseURL = getHubBaseURL()
  return get<PubKeyItem[]>(`${baseURL}/pub_key/list`)
}

export async function deletePubKey(id: number): Promise<void> {
  const baseURL = getHubBaseURL()
  await del(`${baseURL}/pub_key/${id}`)
}
